src/walker.ts

```diff
@@ -0,0 +1,62 @@
+import { PotentialStep, Pipeline, Conditional, Command, CommandStep } from '.';
+import { getAndCacheDependency } from './conditional';
+
+export type Entity = Pipeline | PotentialStep | Command;
+
+async function walkEntity<T extends Entity>(
+  entity: T,
+  nodeCache: Map<any, any>,
+  mutator?: (entity: Entity) => any,
```

Comment:
the mutator function should be its own exported type and most definitely not allow returning `any`. You can also define a default mutator which returns identity.

In fact, what happens if I mutate from a one type of entity to another? E.g. let's say during a walk I get a `Pipeline` but I mutate to a `Command` or `PotentialStep`? I think possibly `Entity` needs to be generic to allow only same-type mutations?

Reply:
Would it be more idiomatic if instead of `<E extends Entity> (entity: E) => E` it was:
```
{
  pipelineFn?: (pipeline: Pipeline) => Pipeline,
  stepFn?: (step: Step) => Step,
  commandFn?: (command: Command) => Command,
}
```

Saves the user from doing their own `instanceof` checks.

Reply:
I thought about doing that, but I thought it was a worse user experience if they needed to pass in 3 different functions to do that vs `instanceof` checks.

Could be wrong about this so happy to defer to this approach if we think it's best

Reply:
I can't figure out a good way to type the MutatorFn via generics, I suspect TS has a limitation on attempting to cast a generic's subtype to the generic, but it could also be my TS-fu being not very good. I've spent a while trying to figure this out, and it may be easier to just use @uri-canva's suggestion of just splitting it into 3 different methods.

For now I've casted it to any here: https://github.com/joscha/buildkite-graph/pull/74/files#diff-b60258e6578a167b89d5bd3b1029e01b72395435edc1ebafbe4e6669900168d5R55, but since the consumer will be writing this method, we definitely shouldn't leave it like this, so consider that line WIP for now

Reply:
Actioned @uri-canva's suggestion of using 3 mutator methods. This does mean the original `walkEntity` method would experience the same issues, so I've split that into 3 different methods as well.